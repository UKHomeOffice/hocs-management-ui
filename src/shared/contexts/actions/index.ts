import types from './types';

export interface Action<T> {
    type: string;
    payload: T;
}

export type ActionFunctionAny<R> = (...args: any[]) => R;

function createAction<Payload>(actionType: string) : ActionFunctionAny<Action<any>> {
    return (payload: Payload) => ({
        payload,
        type: actionType
    });
}

export const unsetError = createAction(types.UNSET_ERROR);

export const clearApiStatus = createAction(types.CLEAR_API_STATUS);
