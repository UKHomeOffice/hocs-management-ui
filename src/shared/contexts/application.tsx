import React, { Component } from 'react';
import types from './actions/types';
import { ContextAction } from './actions/index';
import { ErrorContent } from '../layouts/error';
import Config, { LayoutConfig } from '../models/config';
import ApiStatus from '../models/apiStatus';

export interface ApplicationState {
    apiStatus?: ApiStatus;
    csrf?: string;
    dispatch(action: ContextAction<any>): Promise<any>;
    error?: ErrorContent;
    layout?: LayoutConfig;
    hasRole(role: string): boolean;
    hasOneOfRoles(roles: string[]): boolean;
}

interface ApplicationProps {
    children: React.ReactElement;
    config: Config;
}

const defaultState: ApplicationState = {
    dispatch: () => Promise.resolve(),
    hasRole: (_: string) => { return false; },
    hasOneOfRoles: (_: string[]) => { return false; }
};

export const Context = React.createContext<ApplicationState>(defaultState);

const reducer = (state: ApplicationState, action: ContextAction<any>): ApplicationState => {
    let newState;
    switch (action.type) {

        case types.UNSET_ERROR:
            newState = { ...state, error: undefined };
            break;
        case types.CLEAR_API_STATUS:
            newState = { ...state, apiStatus: undefined };
            break;
        case types.UPDATE_API_STATUS:
            newState = { ...state, apiStatus: { ...action.payload } };
            break;
        default:
            console.warn('Unsupported action');
            newState = state;
    }
    return newState;
};

export class ApplicationProvider extends Component<ApplicationProps, ApplicationState> {

    constructor(props: ApplicationProps) {
        super(props);
        const { config } = props;

        console.log(props)

        this.state = {
            ...props.config,
            apiStatus: undefined,
            dispatch: (action: ContextAction<any>) => {
                try {
                    this.setState(state => reducer(state, action));
                    return Promise.resolve();
                } catch (error) {
                    return Promise.reject(error);
                }
            },
            hasRole: (role) => {
                return config.user?.roles.includes(role) || false;
            },
            hasOneOfRoles: (roles: string[]) => {
                return roles.some(role => config.user?.roles.includes(role)) || false;
            }
        };
    }

    render() {
        const { children } = this.props;
        return (
            <Context.Provider value={this.state}>
                {children}
            </Context.Provider>
        );
    }

}

export const ApplicationConsumer = Context.Consumer;
