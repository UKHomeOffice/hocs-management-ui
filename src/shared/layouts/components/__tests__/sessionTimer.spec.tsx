import React from 'react';
import { render } from '@testing-library/react';
import SessionTimer from '../sessionTimer';
import { ApplicationProvider } from '../../../contexts/application';
import { HelmetProvider } from 'react-helmet-async';
import Modal from 'react-modal';

jest
    .spyOn(Modal, 'setAppElement')
    .mockImplementation(param => console.log(param));
describe('<Component />', () => {});

const config = {
    analytics: undefined,
    csrf: '',
    layout: {
        body: { phaseBanner: { feedback: '', isVisible: true, phase: '' } },
        countDownForSeconds: 5,
        defaultTimeoutSeconds: 10,
        footer: { isVisible: true, links: [] },
        header: {
            isVisible: true,
            service: 'service name',
            serviceLink: '',
        },
    },
    user: {
        roles: []
    }
};

describe('Session timer component', () => {

    it('should render with default props', () => {

        expect(
            render(
                <HelmetProvider>
                    <ApplicationProvider config={config}>
                        <SessionTimer />
                    </ApplicationProvider>
                </HelmetProvider>
            )
        ).toMatchSnapshot();
    });

    it('should render the modal when timing out', () => {
        jest.spyOn(React, 'useContext').mockReturnValue({
            layout: {
                countDownForSeconds: 5,
                defaultTimeoutSeconds: 4,
                header: {
                    service: 'service name'
                }
            }
        });

        expect(
            render(
                <HelmetProvider>
                    <ApplicationProvider config={config}>
                        <SessionTimer />
                    </ApplicationProvider>
                </HelmetProvider>
            )
        ).toMatchSnapshot();
    });
    it('should render the modal when timed out', () => {
        jest.spyOn(React, 'useContext').mockReturnValue({
            layout: {
                countDownForSeconds: 5,
                defaultTimeoutSeconds: 0,
                header: {
                    service: 'service name'
                }
            }
        });

        expect(
            render(
                <HelmetProvider>
                    <ApplicationProvider config={config}>
                        <SessionTimer />
                    </ApplicationProvider>
                </HelmetProvider>
            )
        ).toMatchSnapshot();
    });
});
