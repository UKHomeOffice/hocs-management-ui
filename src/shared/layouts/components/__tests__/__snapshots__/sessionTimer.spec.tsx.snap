// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Session timer component should render the modal when timed out 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="ReactModal__Body--open"
  >
    <div
      data-react-modal-body-trap=""
      style="position: absolute; opacity: 0;"
      tabindex="0"
    />
    <div />
    <div
      data-react-modal-body-trap=""
      style="position: absolute; opacity: 0;"
      tabindex="0"
    />
    <div
      class="ReactModalPortal"
    >
      <div
        class="ReactModal__Overlay"
        style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(189, 189, 189, 0.75); display: flex; z-index: 3;"
      >
        <div
          aria-modal="true"
          class="ReactModal__Content"
          role="dialog"
          style="position: relative; top: 40px; left: 40px; right: 40px; bottom: 40px; background: rgb(255, 255, 255); overflow: auto; border-radius: 0px; outline: none; padding: 20px; margin: auto;"
          tabindex="-1"
        >
          <h2
            class="govuk-heading-l"
          >
            Your session has expired
          </h2>
          <p
            class="govuk-body"
          >
            You'll need to login again.
          </p>
          <button
            class="govuk-button"
          >
            Return to login
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Session timer component should render the modal when timing out 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="ReactModal__Body--open"
  >
    <div
      data-react-modal-body-trap=""
      style="position: absolute; opacity: 0;"
      tabindex="0"
    />
    <div />
    <div
      data-react-modal-body-trap=""
      style="position: absolute; opacity: 0;"
      tabindex="0"
    />
    <div
      class="ReactModalPortal"
    >
      <div
        class="ReactModal__Overlay"
        style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(189, 189, 189, 0.75); display: flex; z-index: 3;"
      >
        <div
          aria-modal="true"
          class="ReactModal__Content"
          role="dialog"
          style="position: relative; top: 40px; left: 40px; right: 40px; bottom: 40px; background: rgb(255, 255, 255); overflow: auto; border-radius: 0px; outline: none; padding: 20px; margin: auto;"
          tabindex="-1"
        >
          <h2
            class="govuk-heading-l"
          >
            Your session will expire in 4 seconds
          </h2>
          <p
            class="govuk-body"
          >
            We won't be able to save what you have done and you'll lose your progress. 
          </p>
          <p
            class="govuk-body"
          >
             Click Continue to extend your session.
          </p>
          <button
            class="govuk-button"
          >
            Continue
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Session timer component should render with default props 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div
      class="ReactModalPortal"
    />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
